src/mds/Locker.cc

```diff
@@ -2991,6 +2992,10 @@ void Locker::_update_cap_fields(CInode *in, int dirty, MClientCaps *m, inode_t *
 	      << " for " << *in << dendl;
       pi->mtime = mtime;
     }
+    if (((dirty & CEPH_CAP_FILE_WR) && change_attr > pi->change_attr) ||
+	((dirty & CEPH_CAP_FILE_EXCL) && change_attr != pi->change_attr)) {
+      pi->change_attr = change_attr;
```

Comment:
Actually, this is a little off. We need the mds to take the latest change_attr seen, regardless of what caps are being returned, as long as some of them are dirty (much like we do with the ctime).

You don't really want to bump it if no one else actually changed it. If we were to just bump it like you suggest, then solo NFS clients may end up with spurious cache invalidations, and that's not a good thing either.

While the logic is somewhat complicated, the saving grace here is that the change_attr can only tell you if _something_ has changed since you last checked it. It's semi-opaque in that you can check to see if it's larger than a previous value, but can't infer anything from the _actual_ value. Thus we can allow a single bump of the change_attr to encompass many changes, as long as it shows as changed when a client queries for it.